### PR TITLE
[Merged by Bors] - Monitoring service api

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

N/A

## Proposed Changes

Adds a client side api for collecting system and process metrics and pushing it to a monitoring service.

### CODE REVIEW

File: beacon_node/src/cli.rs
Comment (paulhauner) [inline beacon_node/src/cli.rs]: Are the `explorer-metrics` and `explorer-address` flags standard between clients? Just out of interest.

File: beacon_node/src/config.rs
Comment (paulhauner) [inline beacon_node/src/config.rs]: I would just return an error with some useful text here, it'll be nicer than a `panic: guaranteed by clap` if we happen to mess up the clap config.

**Review by paulhauner — COMMENTED** [review]
I can see there's still some `TODO` and commented-out code fragments so I assume you're after a birds-eye view to ensure things are on the right track. As such, I haven't done a detailed review.

Does the BN/VC do a HTTP call back to itself to gather the metrics from Prometheus? If so, why not just gather directly from the registry instead of going "outside the binary" to collect via HTTP?

**Comment by pawanjay176** [comment]
> Does the BN/VC do a HTTP call back to itself to gather the metrics from Prometheus? If so, why not just gather directly from the registry instead of going "outside the binary" to collect via HTTP?

Had it this way when BN and VC process metrics were both collected from a BN service. Can do as you suggested now that we have BN and VC separate. This ought to simplify things a lot. Thanks!

File: beacon_node/src/cli.rs
Comment (pawanjay176) [reply beacon_node/src/cli.rs]: Nope that's not standard. Can be anything.

**Comment by paulhauner** [comment]
@pawanjay176 is this ready for review?

File: beacon_node/eth1/src/metrics.rs
Comment (paulhauner) [inline beacon_node/eth1/src/metrics.rs]: Perhaps the comment could be "Set to 1 if connected to an eth1 node, otherwise set to 0"

File: beacon_node/eth1/src/metrics.rs
Comment (paulhauner) [inline beacon_node/eth1/src/metrics.rs]: I guess this is set to 0 if there are no connected fallbacks, but set to 1 if there is a connected fallback? Perhaps we could clarify this comment.

File: beacon_node/src/cli.rs
Comment (paulhauner) [inline beacon_node/src/cli.rs]: > Should enable metrics to get all process data

Is this still relevant?

File: beacon_node/src/cli.rs
Comment (paulhauner) [inline beacon_node/src/cli.rs]:
```suggestion
                .help("Set the endpoint where the beacon node metrics will be sent")
```

File: beacon_node/src/cli.rs
Comment (paulhauner) [inline beacon_node/src/cli.rs]: Thinking about these two flags:

- `--explorer-metrics`: present to enable push metrics.
- `--explorer-endpoint`: specifies address for push metrics (must be supplied if previous flag is supplied.)

Could we simply have one flag, `--explorer-endpoint`?

I'm also a little unsure about exposing it as "explorer", to try and distance it from beaconcha.in and make it more general. Perhaps we could call it `--monitoring-endpoint`? What are your thoughts on that?

File: beacon_node/eth1/src/service.rs
Comment (paulhauner) [inline beacon_node/eth1/src/service.rs:751]: Perhaps we should also do `metrics::set_gauge(&metrics::ETH1_FALLBACK_CONNECTED, 0);` here as well:

https://github.com/sigp/lighthouse/blob/1296a3c02cb27a727b7a19f0de9518ea324c2d76/beacon_node/eth1/src/service.rs#L115

Not sure though, you might have thought about it more than me.

File: validator_client/src/cli.rs
Comment (paulhauner) [inline validator_client/src/cli.rs:187]: As per comments on BN flags.

File: validator_client/src/http_metrics/metrics.rs
Comment (paulhauner) [inline validator_client/src/http_metrics/metrics.rs]: Like the other comment, perhaps we can describe the values of this gauge (and the one above) in more detail?

File: common/eth2/src/lighthouse.rs
Comment (paulhauner) [inline common/eth2/src/lighthouse.rs]: This is a good place for saturating/checked subtraction.

**Comment by paulhauner** [comment]
Flagged as `backwards-incompat` since it changes the output of https://lighthouse-book.sigmaprime.io/api-lighthouse.html#lighthousehealth

File: common/explorer_api/src/gather.rs
Comment (paulhauner) [inline common/explorer_api/src/gather.rs]: Nice, I like how this is `O(n)` by using the `metrics_map`.

File: common/explorer_api/src/gather.rs
Comment (paulhauner) [inline common/explorer_api/src/gather.rs]: I wouldn't block on it, but I reckon that when there's a 3-tuple with a paragraph to explain the values it's a good time for a struct.

File: common/explorer_api/src/gather.rs
Comment (paulhauner) [inline common/explorer_api/src/gather.rs]: Perhaps we could avoid some resizing allocations with this:

```suggestion
    let mut res = serde_json::Map::with_capacity(metrics_map.len());
```

File: common/explorer_api/src/gather.rs
Comment (paulhauner) [inline common/explorer_api/src/gather.rs]: Will a metric that has never received a value return an error here? E.g., if we had a `num_errors_logged` metric and we'd registered it with Prometheus but never had an error occur.

If that's the case, perhaps we should provide default values for each of the `BEACON_PROCESS_METRICS` and `VALIDATOR_PROCESS_METRICS` items?

File: common/explorer_api/src/types.rs
Comment (paulhauner) [inline common/explorer_api/src/types.rs]: TODO

File: common/lighthouse_version/src/lib.rs
Comment (paulhauner) [inline common/lighthouse_version/src/lib.rs]: Is this supposed to be here?

File: validator_client/src/notifier.rs
Comment (paulhauner) [inline validator_client/src/notifier.rs]: I guess this isn't strictly true since we might have a connected fallback but a disconnected primary.

**Review by paulhauner — CHANGES_REQUESTED** [review]
Getting close! I'm liking this new structure! Nothing structural in the comments, mainly nits and questions :)

File: beacon_node/src/cli.rs
Comment (pawanjay176) [reply beacon_node/src/cli.rs]: Agree on having just a single flag that accepts the monitoring endpoint.

I considered `--monitoring-endpoint` but thought that it might potentially cause some confusion with the validator monitoring flags. Agree on making it more general instead of tightly associating it with beaconcha.in.

Comment (pawanjay176) [reply beacon_node/src/cli.rs]: setting it to `--monitoring-endpoint` because not able to come up with a better name :sweat_smile:

File: common/explorer_api/src/gather.rs
Comment (pawanjay176) [reply common/explorer_api/src/gather.rs]: Thanks for your help on this :)

File: common/explorer_api/src/gather.rs
Comment (pawanjay176) [reply common/explorer_api/src/gather.rs]: Haha totally. This got out of hand quickly :sweat_smile:

File: common/explorer_api/src/types.rs
Comment (pawanjay176) [reply common/explorer_api/src/types.rs]: Currently, we only have a version which is a string. 
Not sure if `build` is something standard that we derive from a version, couldn't find anything regarding this

File: common/lighthouse_version/src/lib.rs
Comment (pawanjay176) [reply common/lighthouse_version/src/lib.rs]: Not sure why I added this. will remove.

File: beacon_node/src/cli.rs
Comment (pawanjay176) [reply beacon_node/src/cli.rs]: Nope, not anymore.

File: common/explorer_api/src/gather.rs
Comment (pawanjay176) [reply common/explorer_api/src/gather.rs]: Done!

File: beacon_node/eth1/src/service.rs
Comment (pawanjay176) [reply beacon_node/eth1/src/service.rs:751]: This is tricky because we don't know which node is down. Could be the primary node here as well.
`ETH1_FALLBACK_CONNECTED` right now checks only if the fallback is configured. Have added a comment here
 https://github.com/sigp/lighthouse/blob/845f22f71ef9cf1a057517d7518b919f76345836/beacon_node/eth1/src/metrics.rs#L42-L45

File: validator_client/src/notifier.rs
Comment (pawanjay176) [reply validator_client/src/notifier.rs]: Fixed this here d55da59 and 7c4de06

File: common/lighthouse_version/src/lib.rs
Comment (pawanjay176) [reply common/lighthouse_version/src/lib.rs]: Actually, using this to return the `client_version` parameter. Forgot to update it.

File: beacon_node/src/cli.rs
Comment (paulhauner) [inline beacon_node/src/cli.rs]:
```suggestion
                This flag sets the endpoint where the beacon node metrics will be sent. ")
```

Also, could we please append a sentence like:

> This will send information to a remote sever which may identify and associate your validators, IP address and other personal information. Always use a HTTPS connection and never provide an untrusted URL.

File: beacon_node/src/cli.rs
Comment (paulhauner) [reply beacon_node/src/cli.rs]: Nice, I'm down with `monitoring-endpoint`!

File: common/lighthouse_version/src/lib.rs
Comment (paulhauner) [inline common/lighthouse_version/src/lib.rs]: Instead of defining a new variable, could we leverage [`lighthouse_version::VERSION`](https://github.com/sigp/lighthouse/blob/3a24ca5f14c6e6d6612fd43eca82aa0c1e6aba16/common/lighthouse_version/src/lib.rs#L11)?

It takes the form of `Lighthouse/vx.x.x-<commit>`. Ideally, we could pass that to the monitoring server. If they specifically want `x.x.x`, we should be able to extract them using regex.

At the moment we have a single-source-of-truth inside Lighthouse for that version string (apart from the `Cargo.toml`) and it would be nice to keep it that way, if we can :D

File: common/explorer_api/src/types.rs
Comment (paulhauner) [reply common/explorer_api/src/types.rs]: Oh yeah, I don't think we have anything like this..

File: common/monitoring_api/src/types.rs
Comment (paulhauner) [inline common/monitoring_api/src/types.rs]:
```suggestion
    0
```

Let's just set this to `0` and forget about it, we don't use build numbers. What do you reckon?

File: common/monitoring_api/src/types.rs
Comment (paulhauner) [inline common/monitoring_api/src/types.rs]:
```suggestion
/// Lighthouse does not support build numbers, this is effectively a null-value.
```

File: validator_client/src/cli.rs
Comment (paulhauner) [inline validator_client/src/cli.rs]: Can we also add that warning text I suggested for the BN, please?

**Review by paulhauner — COMMENTED** [review]
Looking good, only a few tiny things then we're good to go!

EDIT: It seems like Github has spread my comments around pretty weirdly... there's two more comments above this one. Sorry about that.

File: common/lighthouse_version/src/lib.rs
Comment (pawanjay176) [reply common/lighthouse_version/src/lib.rs]: Agree, was pure laziness on my part. Will fix this :)

File: common/monitoring_api/src/types.rs
Comment (pawanjay176) [reply common/monitoring_api/src/types.rs]: SGTM.

File: validator_client/src/cli.rs
Comment (pawanjay176) [reply validator_client/src/cli.rs]: Done!

**Comment by paulhauner** [comment]
Sorry @pawanjay176, it looks like some merge conflicts appeared.

**Comment by pawanjay176** [comment]
Fixed!

File: common/monitoring_api/src/types.rs
Comment (paulhauner) [inline common/monitoring_api/src/types.rs]:
```suggestion
        let misc_os = health.misc_os.get(0..3).unwrap_or("???").to_string();
```

Not sure if this will compile, but hopefully you get the idea.

File: common/monitoring_api/src/lib.rs
Comment (paulhauner) [inline common/monitoring_api/src/lib.rs]:
```suggestion
    pub fn get_beacon_metrics(&self) -> Result<MonitoringMetrics, Error> {
```

File: common/monitoring_api/src/lib.rs
Comment (paulhauner) [inline common/monitoring_api/src/lib.rs]:
```suggestion
    pub fn get_validator_metrics(&self) -> Result<MonitoringMetrics, Error> {
```

File: common/monitoring_api/src/lib.rs
Comment (paulhauner) [inline common/monitoring_api/src/lib.rs]:
```suggestion
    pub fn get_system_metrics(&self) -> Result<MonitoringMetrics, Error> {
```

File: common/monitoring_api/src/types.rs
Comment (paulhauner) [inline common/monitoring_api/src/types.rs]:
```suggestion
        // Export format uses 3 letter os names
```

**Review by paulhauner — CHANGES_REQUESTED** [review]
Only a few tiny nitpicks left! We'll be good to go after they're resolved :)

**Review by paulhauner — APPROVED** [review]
Awesome, LGTM! :rocket: 

bors r+

**Comment by paulhauner** [comment]
bors retry

**Pull request closed without merging**
